### PR TITLE
Fixing bug for addUserRecord

### PR DESCRIPTION
**Pull request by aakkem**
*Issue #, if available:*
Support for CAA

*Description of changes:*

Adding the implementation for addUserRecord 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java [hunk 1]
```diff
@@ -668,6 +668,72 @@ public ListenableFuture<UserRecordResult> addUserRecord(String stream, String pa
...
+     * @param partitionKey
+     *            Partition key. Length must be at least one, and at most 256
+     *            (inclusive).
+     * @param explicitHashKey
+     *            The hash value used to explicitly determine the shard the data
+     *            record is assigned to by overriding the partition key hash.
+     *            Must be a valid string representation of a positive integer
+     *            with value between 0 and <code>2^128 - 1</code> (inclusive).
+     * @param data
+     *            Binary data of the record. Maximum size 1MiB.
```
Comment (shanmsac) [inline java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java:722]: Add @param for `schema` parameter

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java [hunk 2]
```diff
@@ -668,6 +668,72 @@ public ListenableFuture<UserRecordResult> addUserRecord(String stream, String pa
...
+     *             if the child process is dead
+     * @see ListenableFuture
+     * @see UserRecordResult
+     * @see KinesisProducerConfiguration#setRecordTtl(long)
+     * @see UserRecordFailedException
+     */
+    @Override
+    public ListenableFuture<UserRecordResult> addUserRecord(String stream, String partitionKey, String explicitHashKey, ByteBuffer data, Schema schema) {
+        return addUserRecord(stream, partitionKey, explicitHashKey, data, null, schema);
+    }
```
Comment (shanmsac) [inline java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java:736]: new line

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/UserRecord.java [hunk 3]
```diff
@@ -69,6 +69,13 @@ public UserRecord(String streamName, String partitionKey, String explicitHashKey
         this.data = data;
     }
 
+    public UserRecord(String streamName, String partitionKey, String explicitHashKey, ByteBuffer data, Schema schema) {
+        this.streamName = streamName;
+        this.partitionKey = partitionKey;
+        this.explicitHashKey = explicitHashKey;
+        this.data = data;
+        this.schema = schema;
+    }
```
Comment (shanmsac) [inline java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/UserRecord.java:78]: new line

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java [hunk 4]
```diff
@@ -668,6 +668,72 @@ public ListenableFuture<UserRecordResult> addUserRecord(String stream, String pa
...
     @Override
     public ListenableFuture<UserRecordResult> addUserRecord(String stream, String partitionKey, String explicitHashKey, ByteBuffer data, String streamARN, Schema schema) {
```
Comment (shanmsac) [inline java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java:738]: Please add Javadoc comment for this function with all parameters in order

**Review by shanmsac — CHANGES_REQUESTED** [review]
minor comments

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java (hunk 1 above)
Comment (aakkem) [reply java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java:722]: Addressed the comments.

**Review by shanmsac — APPROVED** [review]
lgtm

**Comment by kabennett** [comment]
When are you going to make the 0.15.10 release available in Maven Central?

**Comment by aakkem** [comment]
This should be available by the end of the week.

**Comment by kabennett** [comment]
Thanks. Do you know if this release will fix the errors I am encountering as documented in stackoverflow [here](https://stackoverflow.com/questions/77788780/kpl-errors-encountered-after-upgrading-spring-cloud-stream-binder-for-kinesis-to)?

**Comment by kabennett** [comment]
Hey, @aakkem, are you going to release 0.15.10 to Maven Central?

**Comment by ofeki-neosec** [comment]
Hey @aakkem, is there an estimation on when this will be released to Maven Central?

**Comment by kabennett** [comment]
@ofeki-neosec take a look at[ this stackoverflow link](https://stackoverflow.com/questions/77788780/kpl-errors-encountered-after-upgrading-spring-cloud-stream-binder-for-kinesis-to) I posted a few weeks ago. I updated that post with a comment you may find helpful. Are you having the same issue?

**Comment by joshuabaird** [comment]
I believe this bug is also causing segfaults when using IRSA with EKS to provide credentials.  Reverting to v0.15.8 fixes the issue.  See https://github.com/awslabs/amazon-kinesis-producer/issues/558 for details.

**Comment by svenwilliger** [comment]
@aakkem could you please release it to Maven Central now ?

**Pull request merged**